### DISCUSSION TITLE
Order-line item price computation with 0 decimals and tax inclusive price settings.

### DISCUSSION THREAD

**Discussion by Sidsector9** [woocommerce/woocommerce#46126] [Checkout Flow]
We are facing an edge case problem while rendering the filtered order-line item prices.

### The problem: **Aggregation bias caused by error propagation due to rounding with 0 decimal numbers.**

We are using the [WooCommerce Deposits](https://woo.com/products/woocommerce-deposits/) plugin.

---

In the following scenario where:

1. A product has a 50% deposit plan applied
3. The **number of decimal** setting is set to `0`

On the block Cart page, **the order-line item of a product with deposit applied is incorrect.**

The Deposits extension makes use of the `rest_request_after_callbacks` filter hook to filter the order-line item subtotal. For example like the following:

<img width="829" alt="Screenshot 2024-04-02 at 7 09 28 PM" src="https://github.com/woocommerce/woocommerce/assets/17757960/65febf84-8907-4037-b344-4f68697daa6c">

If you observe the image above, the sum of the prices in red is higher than the price in green.

In the callback for `rest_request_after_callbacks`, Deposit sets the prices in red as following:

```php
$deposit_amount_including_tax = wc_get_price_including_tax( 
 	$cart_item['data'], 
 	array( 
 		'qty'   => $cart_item['quantity'], 
 		'price' => $cart_item_deposit_amount, 
 	) 
 ); 
  
$data['items'][ $key ]['totals']->line_subtotal     = $money_formatter->format( $deposit_amount_excluding_tax ); 
$data['items'][ $key ]['totals']->line_subtotal_tax = $money_formatter->format( $deposit_amount_including_tax - $deposit_amount_excluding_tax ); 
```

Because the decimal setting is set to `0`, `$money_formatter->format()` rounds up the prices if the decimal is `>= 0.5` and rounds down if the decimal is `< 0.5` . So this is what is happening:

![pre-orders-586 drawio](https://github.com/woocommerce/woocommerce/assets/17757960/877b9b7b-6112-48f6-be5d-1f6bc9284c1a)

At the same time, `$data['items'][ $key ]['totals']->line_subtotal` and `$data['items'][ $key ]['totals']->line_subtotal_tax` cannot be set to floating numbers as only the integer part is considered by Woo Core on the client side:

https://github.com/woocommerce/woocommerce/blob/7fe00de7278cbc0d7806e1733b8df7320feef933/plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/cart-line-items-table/cart-line-item-row.tsx#L152-L155

Would you have any advice to best handle a situation like this?

**Comment by senadir** [comment]
Would doing ` $money_formatter->format( $deposit_amount_including_tax ) -  $money_formatter->format(  $deposit_amount_excluding_tax );` works?

**Comment by Sidsector9** [comment]
@senadir Thanks for looking into this. Your solution will work for the prices given in this example but will fail when the product prices are $7.75 or $13.5.

**Reply by senadir** [reply]
The product price can never be 13.5 because WooCommerce will round it at save.

**Reply by Sidsector9** [reply]
@senadir apologies for the delay, I was on leave.

> The product price can never be 13.5 because WooCommerce will round it at save.

During the computation of total it does not consider the rounded values. I have tried with multiple values and found the issue to be consistent.

**Comment by Sidsector9** [comment]
Closing as this has been resolved with an "adjustment" fix.
